File: data-structures-algorithms-coursera/course1-algorithmic-toolbox/week3_greedy_algorithms/2_maximum_value_of_the_loot/fractional_knapsack.py
# Input: The first line of the input contains the number n of items 
# and the capacity W of a knapsack. The next n lines define the values 
# and weights of the items. The i-th line contains integers v_i and w_i — the
# value and the weight of i-th item, respectively.
# Output: the maximal value of fractions of items that fit into the knapsack. 
# The absolute value of the difference between the answer of your program and 
# the optimal value should be at most 1e−3. To ensure this, output your answer 
# with at least four digits after the decimal point (otherwise your answer, 
# while being computed correctly, can turn out to be wrong because of rounding issues).


def most_expensive_index(weights, values) -> int:
    """
    Get the index of the most expensive item
    """
    max_index = 0
    max_cost = -1
    for i in range(len(weights)):
        cost: float = values[i] / weights[i]
        if cost > max_cost:
            max_cost = cost
            max_index = i

    return max_index


def get_optimal_value(capacity, weights, values):
    """
    Strategy: Get the most expensive item first
    """
    # base case
    if capacity == 0 or len(weights) == 0:
        return 0
    # recursive case
    m = most_expensive_index(weights, values)
    amount_taken = min(capacity, weights[m])
    value: float = float(amount_taken * float(values[m] / weights[m]))
    # remove the taken item 
    capacity = capacity - amount_taken
    del weights[m]
    del values[m]

    return value + get_optimal_value(capacity, weights, values)   


if __name__ == "__main__":
    # get user input
    data = list(map(int, input().split()))
    n, capacity = data
    values = []
    weights = []
    for i in range(n):
        v, w = list(map(int, input().split()))
        values.append(v)
        weights.append(w)
    # get the optimal 
    opt_value = get_optimal_value(capacity, weights, values)
    print("{:.10f}".format(opt_value))

File: data-structures-algorithms-coursera/course1-algorithmic-toolbox/week3_greedy_algorithms/2_maximum_value_of_the_loot/fractional_knapsack.cpp
// Input: The first line of the input contains the number n of items 
// and the capacity W of a knapsack. The next n lines define the values 
// and weights of the items. The i-th line contains integers v_i and w_i — the
// value and the weight of i-th item, respectively.
// Output: the maximal value of fractions of items that fit into the knapsack. 
// The absolute value of the difference between the answer of your program and 
// the optimal value should be at most 1e−3. To ensure this, output your answer 
// with at least four digits after the decimal point (otherwise your answer, 
// while being computed correctly, can turn out to be wrong because of rounding issues).


#include <iostream>
#include <vector>
#include <algorithm>
#include <iomanip>

using std::vector;

int most_expensive_index(vector<int> weights, vector<int> values) {
  // return the index of the most expensive item
  int max_index = 0;
  double max_value = -1;
  for (int i = 0; i < weights.size(); ++i) {
    double cost = (double) values[i] / weights[i];
    if (cost > max_value) {
      max_index = i;
      max_value = cost;
    }
  }
  return max_index;
}

double get_optimal_value(int capacity, vector<int> weights, vector<int> values) {
  // idea: get the item with maximum value each time until we run out of capacity
  //       or we have got all the items

  // base case
  if (capacity == 0 || weights.size() == 0) {
    return 0.0;
  }

  // recursive case
  // the index of the most expensive item
  // int m = std::max_element(values.begin(), values.end()) - values.begin();
  int m = most_expensive_index(weights, values);
  int amount_taken = std::min(weights[m], capacity);
  double value = (double) values[m] / weights[m];
  value = (double) value * amount_taken;
  // std::cout << std::fixed << value << " \n";
  // remove the taken items
  capacity = capacity - amount_taken;
  weights.erase(weights.begin() + m);
  values.erase(values.begin() + m);

  return value + get_optimal_value(capacity, weights, values);
}

int main() {
  // get user input
  int n;
  int capacity;
  std::cin >> n >> capacity;
  vector<int> values(n);
  vector<int> weights(n);
  for (int i = 0; i < n; i++) {
    std::cin >> values[i] >> weights[i];
  }

  std::cout << std::fixed;
  std::cout << std::setprecision(4);
  double optimal_value = get_optimal_value(capacity, weights, values);
  std::cout << optimal_value << std::endl;

  return 0;
}
